ext/pdo/pdo_stmt.c

```diff
@@ -1108,44 +1046,38 @@ PHP_METHOD(PDOStatement, fetchObject)
 {
 	zend_class_entry *ce = NULL;
 	zend_class_entry *old_ce;
-	zval old_ctor_args, *ctor_args = NULL;
-	int old_arg_count;
+	HashTable *old_ctor_args, *ctor_args = NULL;
 
 	ZEND_PARSE_PARAMETERS_START(0, 2)
 		Z_PARAM_OPTIONAL
 		Z_PARAM_CLASS_OR_NULL(ce)
-		Z_PARAM_ARRAY(ctor_args)
+		Z_PARAM_ARRAY_HT(ctor_args)
 	ZEND_PARSE_PARAMETERS_END();
 
 	PHP_STMT_GET_OBJ;
 	PDO_STMT_CLEAR_ERR();
 
 	old_ce = stmt->fetch.cls.ce;
-	ZVAL_COPY_VALUE(&old_ctor_args, &stmt->fetch.cls.ctor_args);
-	old_arg_count = stmt->fetch.cls.fci.param_count;
-
-	do_fetch_opt_finish(stmt, 0);
-
-	if (ctor_args && zend_hash_num_elements(Z_ARRVAL_P(ctor_args))) {
-		ZVAL_ARR(&stmt->fetch.cls.ctor_args, zend_array_dup(Z_ARRVAL_P(ctor_args)));
-	} else {
-		ZVAL_UNDEF(&stmt->fetch.cls.ctor_args);
+	old_ctor_args = stmt->fetch.cls.ctor_args;
+	stmt->fetch.cls.ctor_args = NULL;
+	if (ce == NULL) {
+		ce = zend_standard_class_def;
 	}
-	if (ce) {
-		stmt->fetch.cls.ce = ce;
-	} else {
-		stmt->fetch.cls.ce = zend_standard_class_def;
+
+	if (ctor_args && zend_hash_num_elements(ctor_args) && ce->constructor == NULL) {
+		zend_argument_value_error(2, "must be empty when class provided in argument #1 ($class) does not have a constructor");
```

Comment:
You don't restore `stmt->fetch.cls.ctor_args` on error (set to NULL by `stmt->fetch.cls.ctor_args = NULL;`), so you can leak that value.
In fact, this check should probably be done as early as possible before any modification.